src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,118 +291,188 @@ class HandshakeALotClosure : public HandshakeClosure {
+  if ((SelfDestructTimer != 0) && !VMError::is_error_reported() &&
+      (os::elapsedTime() > (double)SelfDestructTimer * 60.0)) {
+    tty->print_cr("VM self-destructed");
+    exit(-1);
+  }
+}
 
-    //
-    // Execute VM operation
-    //
-    { HandleMark hm(VMThread::vm_thread());
+void VMThread::inner_execute(VM_Operation* op) {
+  Thread* current = Thread::current();
+  assert(current->is_VM_thread(), "must be a VM thread");
+
+  VM_Operation* prev_vm_operation = vm_operation();
+  if (prev_vm_operation != NULL) {
+    // Check the VM operation allows nested VM operation.
+    // This normally not the case, e.g., the compiler
+    // does not allow nested scavenges or compiles.
+    if (!prev_vm_operation->allow_nested_vm_operations()) {
+      fatal("Nested VM operation %s requested by operation %s",
+            op->name(), vm_operation()->name());
+    }
+    op->set_calling_thread(prev_vm_operation->calling_thread());
+  }
+
+  _cur_vm_operation = op;
 
-      EventMark em("Executing VM operation: %s", vm_operation()->name());
-      assert(_cur_vm_operation != NULL, "we should have found an operation to execute");
+  HandleMark hm(VMThread::vm_thread());
+  EventMark em("Executing %s VM operation: %s", prev_vm_operation ? "nested" : "", op->name());
 
-      // If we are at a safepoint we will evaluate all the operations that
-      // follow that also require a safepoint
-      if (_cur_vm_operation->evaluate_at_safepoint()) {
-        log_debug(vmthread)("Evaluating safepoint VM operation: %s", _cur_vm_operation->name());
+  // If we are at a safepoint we will evaluate all the operations that
+  // follow that also require a safepoint
+  log_debug(vmthread)("Evaluating %s %s VM operation: %s",
+                       prev_vm_operation ? "nested" : "",
+                      _cur_vm_operation->evaluate_at_safepoint() ? "safepoint" : "non-safepoint",
+                      _cur_vm_operation->name());
+
+  bool end_safepoint = false;
+  if (_cur_vm_operation->evaluate_at_safepoint() &&
+      !SafepointSynchronize::is_at_safepoint()) {
+    SafepointSynchronize::begin();
+    if (_timeout_task != NULL) {
+      _timeout_task->arm();
+    }
+    end_safepoint = true;
+  }
 
-        SafepointSynchronize::begin();
+  evaluate_operation(_cur_vm_operation);
 
-        if (_timeout_task != NULL) {
-          _timeout_task->arm();
-        }
+  if (end_safepoint) {
+    if (_timeout_task != NULL) {
+      _timeout_task->disarm();
+    }
+    SafepointSynchronize::end();
+  }
 
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  _cur_vm_operation = prev_vm_operation;
+}
 
-        if (_timeout_task != NULL) {
-          _timeout_task->disarm();
-        }
+void VMThread::wait_for_operation() {
+  MonitorLocker mu_queue(VMOperation_lock, Mutex::_no_safepoint_check_flag);
 
-        // Complete safepoint synchronization
-        SafepointSynchronize::end();
+  // Notify previous op done
+  mu_queue.notify();
```

Comment:
So, `notify`, not `notifyAll` here? We sure we want to bet who -- of all current waiters -- gets the notification?

Reply:
Will be changed, due to above issue.

Reply:
Please go over the changes in fceebea and see if you agree that it is correct :)
Passes t1-t8.